### PR TITLE
Context Menu Feature

### LINKED ISSUE #389: Feat: Create context menu for <EventBlock> component
We want to allow quick actions on events without opening the event editor.

The planned strategy is to have a custom context menu which appears when right-clicking the event.

Development tasks:
- Create a reusable event context menu component
- create logic to show menu when right-clicking on <EventBlock>
- show a single entry in the context menu which lists the clicked event ID

We will use this as a springboard to allow quick actions on selection and, eventually multiple selections.

### PR DESCRIPTION
**Pull request by asharonbaltazar**
closes #389 

#### Changelog
- create `ContextMenuProvider` component responsible for determining look, state, position, and options of context menu. 
- wrap `App.tsx` with `ContextMenuProvider` 
- create `useContextMenu` hook to access said context. It receives an `Option[]` arg and returns an event handler that should be passed to a `onContextMenu`. (However, the returned event handler can be used for any of the element's event callbacks, but perhaps this flexibility is welcome.)

The above implementation does have an issue with positioning. I spent a few hours trying to pinpoint this bug, but I believe it comes down to how the elements on the page are positioned. I'd love some help with this!

https://github.com/cpvalente/ontime/assets/58940073/fce9cecf-5cad-4dcb-9dab-963df8285673

Happy to discuss further feedback!

### CODE REVIEW

File: apps/client/src/App.tsx [hunk 1]
```diff
@@ -49,16 +50,18 @@ function App() {
...
-            <div className='App'>
-              <ErrorBoundary>
-                <TranslationProvider>
-                  <AppRouter />
-                </TranslationProvider>
-              </ErrorBoundary>
-              <ReactQueryDevtools initialIsOpen={false} />
-            </div>
-          </BrowserRouter>
+          <ContextMenuProvider>
```
Comment (cpvalente) [inline apps/client/src/App.tsx:53]: We don't need to wrap the entire app in the context menu.
Wrapping the element with `data-testid:"event-editor"` in `Editor.tsx` would be enough

Although your solution would have generally been correct, in ontime's case, the views are not interactive and often run in devices like TVs. No need to add code there

**Review by cpvalente — CHANGES_REQUESTED** [review]
Honestly, the implementation here is very clean, and I really like your approach.
I have nothing to comment on other than a minor suggestion on the context placement

I also looked into the positioning issue you mentioned.
As far as I could test, the IDs always resolved correctly. The only thing I saw in your video was that the context menu does not follow the screen. This is common behaviour for a non-browser context menu and not an issue for us

We could potentially attach an event listener to the scroll and close the menu. But I wonder whether it would be worth it 

The render time for the context menu is not great (around 20ms on average) but I guess this is due to Chakra UI. You did great into lazy loading the menu so that there is no initial performance hit

Altogether this is great work and I am keen to integrating it into the app.

We would need to add some actual behaviour into the context menu, but we should do that separately.

I look forwards to seeing more code from you

File: apps/client/src/App.tsx (hunk 1 above)
Comment (asharonbaltazar) [reply apps/client/src/App.tsx]: Addressed with https://github.com/cpvalente/ontime/pull/410/commits/c9f0b08e5da9e4f0e1c297758c0c10b805f43740

**Comment by asharonbaltazar** [comment]
> We could potentially attach an event listener to the scroll and close the menu. But I wonder whether it would be worth it

We could emulate the browser behavior and prevent any scrolling when the context menu is open

**Comment by cpvalente** [comment]
> > We could potentially attach an event listener to the scroll and close the menu. But I wonder whether it would be worth it
> 
> We could emulate the browser behavior and prevent any scrolling when the context menu is open

That is a better idea, thank you for looking out.
Would you like to implement?

**Comment by asharonbaltazar** [comment]
>  Wrapping the element with data-testid:"event-editor" in Editor.tsx would be enough

BTW, `/rundown` doesn't work as it's not sharing the `ContextMenuProvider` context

**Comment by cpvalente** [comment]
> If you're able to give it a test, that'd be much appreciated!

Great work, it looks perfect to me.
Reverting to App.tsx is probably the best for now.

**Comment by asharonbaltazar** [comment]
>  Reverting to App.tsx is probably the best for now.
Reverted

**Comment by cpvalente** [comment]
Thank you again @asharonbaltazar 
I will merge this as is


Could I interest you on another issue? Perhaps something more challenging?

**Comment by asharonbaltazar** [comment]
> Could I interest you on another issue? Perhaps something more challenging?

Let's give it a go. But depending on how difficult it is, I can't promise anything in terms of a timeline.

**Review by cpvalente — APPROVED** [review]

**Comment by cpvalente** [comment]
How about one of these?

They are still good in the way that you don't have to know much about the app to be able to do it.

https://github.com/cpvalente/ontime/issues/388 <---- React only
https://github.com/cpvalente/ontime/issues/403 <---- React + node

**Comment by asharonbaltazar** [comment]
#388 looks tame enough. I'll start with that. Can you assign the issue to me?

**Comment by cpvalente** [comment]
Since you are not a collaborator, I am unable to assign it to you without you being involved in the issue. Would you mind dropping a comment in the ticket?

**Comment by cpvalente** [comment]
Sure, if you are up for a challenge, this is it https://github.com/cpvalente/ontime/issues/358
Otherwise we can have a chat, next big task in my list is https://github.com/cpvalente/ontime/issues/100#issuecomment-1556267435

**Comment by asharonbaltazar** [comment]
Alrighty, I'll keep your comment as a bookmark. I'll work on the form ticket, and once finished, I'll check back here. 
Appreciate the quick follow-ups!

**Pull request merged**